respy/state_space.py

```diff
@@ -0,0 +1,430 @@
+        return np.array([len(range(i.start, i.stop)) for i in self.slices_by_periods])
+
+    @property
+    def num_states(self):
+        """Get the total number states in the state space."""
+        return self.states.shape[0]
+
+    def update_systematic_rewards(self, optim_paras, options):
+        self.wages, self.nonpec = _create_reward_components(
+            self.states[:, -1], self.covariates, optim_paras, options
+        )
+
+    def get_attribute_from_period(self, attr, period):
+        """Get an attribute of the state space sliced to a given period.
+
+        Parameters
+        ----------
+        attr : str
+            String of attribute name, e.g. ``"states"``.
+        period : int
+            Attribute is retrieved from this period.
+
+        """
+        if attr == "covariates":
+            raise AttributeError("Attribute covariates cannot be retrieved by periods.")
+        else:
+            pass
+
+        try:
+            attribute = getattr(self, attr)
+        except AttributeError as e:
+            raise AttributeError(f"StateSpace has no attribute {attr}.").with_traceback(
+                e.__traceback__
+            )
+
+        try:
+            indices = self.slices_by_periods[period]
+        except IndexError as e:
+            raise IndexError(f"StateSpace has no period {period}.").with_traceback(
+                e.__traceback__
+            )
+
+        return attribute[indices]
+
+    def _create_slices_by_periods(self, n_periods):
+        """Create slices to index all attributes in a given period.
+
+        It is important that the returned objects are not fancy indices. Fancy indexing
+        results in copies of array which decrease performance and raise memory usage.
+
+        """
+        self.slices_by_periods = []
+        for i in range(n_periods):
+            idx_start, idx_end = np.where(self.states[:, 0] == i)[0][[0, -1]]
+            self.slices_by_periods.append(slice(idx_start, idx_end + 1))
+
+
+def _create_state_space(options, n_types):
+    """Create the state space.
+
+    This process involves two steps. First, the core state space is created which
+    abstracts from different initial experiences and instead uses the minimum initial
+    experience per choice.
+
+    Secondly, the state space is adjusted by all combinations of initial experiences and
+    also filtered, excluding invalid states.
+
+    """
+    data = _create_core_state_space(options, n_types)
+
+    exp_cols = [f"exp_{sector}" for sector in options["choices_w_exp"]]
+
+    df = pd.DataFrame.from_records(
+        data, columns=["period"] + exp_cols + ["lagged_choice", "type"]
+    )
+
+    codes_to_choices = {i: sec for i, sec in enumerate(options["choices"])}
+    df.lagged_choice = df.lagged_choice.astype("category").cat.rename_categories(
+        codes_to_choices
+    )
+
+    df = _adjust_state_space_with_different_starting_experiences(df, options)
+
```

Comment:
I think it would be nicer to call `_filter_state_space` here in the top level function and not in `_adjust_state_space_with ...`. Is that possible?

Reply:
I agree. We can apply the filters directly on the core state space. Then, we also do not have to adjust for initial experiences which makes the formulas easier, but then we need to document that they apply only to additional experience and not total experiences.